### DISCUSSION TITLE
Add Picture-in-Picture button

### DISCUSSION THREAD

**Discussion by XCanG** [tomasklaen/uosc#494] [Ideas]
I have an suggestion to add this sort of button, there is really not much that probably need to be done:

1. Remember current position and size.
2. Resize it to small window in a bottom-right corner.
3. Apply always on top.
4. Optional: hide most of other UI

Clicking button again will restore previous state.

I think it would be really useful feature. I, myself sometimes watching steams in mpv and while I can do this manually with a bunch of hotkeys, this process always take some time and I couldn't revert it instantly as well. I wish to have this sort of convenient button.

**Comment by tomasklaen** [comment]
We can't read or set window position or size. And this is more of a job for desktop environment imo. I'm personally using autohotkey to position window to a specific place:

```ahk
+PrintScreen::
	if WinExist("ahk_exe mpv.exe") {
		WinGetPos,,,W,H
		WinMove,,, 2560-620, 40, 620, % 620 * (H/W)
	}
	return
```

And I have an old **autotop** script which enables on top only when media is playing, or is in fullscreen/maximized mode:

```lua
local options = {enabled = false}
require('mp.options').read_options(options, 'autotop')

function decide_on_top()
	local is_fullscreen = mp.get_property_native('fullscreen') == true
	local is_paused = mp.get_property_native('pause') == true
	mp.set_property_native('ontop', is_fullscreen or (options.enabled and not is_paused))
end

function set_enabled(value)
	options.enabled = value
	mp.commandv('script-message-to', 'uosc', 'set', 'autotop', options.enabled and 'yes' or 'no')
	decide_on_top()
end

mp.observe_property('pause', 'bool', decide_on_top)
mp.observe_property('fullscreen', 'bool', decide_on_top)

mp.add_key_binding(nil, 'toggle', function() set_enabled(not options.enabled) end)

-- uosc integration

mp.register_script_message('set', function(name, value)
	if name == 'autotop' then set_enabled(value == 'yes') end
end)

-- initialization

mp.add_timeout(0.5, function() set_enabled(options.enabled) end)
```

Which I add to controls bar with:

```
toggle:push_pin:autotop@autotop?Auto on top
```

I then have another autohotkey script which converts global PrintScreen to `f` and sends it to mpv player, which fullscreens the window. That causes the autotop script to also on-top it. Useful to temporarily fullscreen a video on the side if something interesting is going on without loosing focus of current app you're using.

**Reply by XCanG** [reply]
Hm, I'm not that familiar with available functionality. It is sad then, that this is not a thing. If making ahk script, then need to add a little more stuff. Window need to remember position and set always on top. Anyway, this would be beyond original suggestion.

**Comment by hooke007** [comment]
mpv doesn't use any OS's PIP api. I think the request should be firstly posted to mpv instead.

**Reply by XCanG** [reply]
When I suggested it here I was not referring to anything related to that.
I know that there is `T cycle ontop`
But I didn't look if there are any windows resizing features. However I know that there some settings for `mpv.conf`, plus I know that there is also keybindings to shift video and resize, so I assumed that there is functionality already. I suspected it is simple 3 commands: enable always-on-top, move, resize. But from other post message, there isn't enough api and it even not possible, so I didn't know about that.

**Comment by mrfragger** [comment]
I have this and works fine for Mac...guessing it'd work for Linux & Windows too
input.conf
Alt+p          cycle ontop; cycle-values geometry '100%:100%' '0%:0%' ;  cycle-values autofit '33%x66%' '100%x100%' #! [Video/Image] > Picture in Picture toggle

Problem is if user presses Alt+p to enter PIP mode puts bottom screen and then instead uses mouse to double click title bar to maximize screen it'll still be in STAY ON TOP mode.  If one just presses Alt+p again it returns to maximized widow and STAY ON TOP mode is off which is perfect.

So I'm trying to create a button in uosc.conf in the controls section but think it can't stack multiple commands.  So looks like I should probably replace fullscreen button to stay on top so one can easily toggle that instead.  Just trying to replicate IINA PIP, Youtube PIP, and Firefox/Librewolf PIP modes. The main difference is all three of those have a button that restores previous state of window.

Tab Playlist,next?> or Enter Next,last?End Last,gap,command:fullscreen:cycle ontop; cycle-values geometry '100%:100%' '0%:0%' ;  cycle-values autofit '33%x66%' '100%x100%'?F stay on top

so gonna have to resort to using this
command:fullscreen:cycle ontop?Ctrl+T stay on top

<img width="1433" alt="image" src="https://github.com/user-attachments/assets/34b6dec0-f856-4c31-b772-7963ac11b4f3" />
